### ISSUE TITLE
Use case: video door entry system

### ISSUE THREAD

**Issue by pgorod** [BelledonneCommunications/linphone-android#1832]
Hi! Happy Linphone user here!

We use Linphone exclusively to answer a door that has a VoIP video camera (2N Helios). In this use case, we would value seeing less video being blocked by other things on screen. And some of the repetitive tasks can be improved. These are suggestions, not really bug reports.

## **1. Video previews while the phone is ringing**

Some video calls allow "video preview", which is always useful, but especially in this case! I can see who's ringing the door even before I answer. 

But in recent versions I see a big overlay blocking the view. This overlay reads "PA", I think that comes from the caller id of the incoming call (`Porta Abrir 7*` which is just a way I have of putting in a reminder of the code to open the door).

![image](https://user-images.githubusercontent.com/15945027/213714442-8c4454c9-4f7b-4ed5-b81d-6b0e11dc3f4c.png)

**To Reproduce**

Get a SIP account that makes video calls
Configure Linphone account and enable video previews for the incoming calls
Make a video call to Linphone app
See the overlay with letters representing the caller id, blocking the video view

**Expected behavior**

I would like to see a full view of the video, with nothing to block it.

## **2. Too many taps to get to the DTMF codes**

To trigger the command to open the door lock I have to put in a "7*". This requires the following steps:

![image](https://user-images.githubusercontent.com/15945027/213729912-88bed2ed-672d-426b-b849-8ebbf6c47d96.png)

After I tap the `...` on the bottom, I have to select `Numpad`:

![image](https://user-images.githubusercontent.com/15945027/213730224-e7c5977a-b350-43a5-a5f8-0239f496a665.png)

... and then I have to tap `7*` to open the door lock.

![image](https://user-images.githubusercontent.com/15945027/213730617-586ffe46-d410-4d50-a716-13e192ff2ce3.png)

The previous Linphone design had one less tap, and when you have to do this 40 times in an afternoon, like I do, you notice those small things 😩 

**Suggestion for improvement**

I think that Grandstream video phones have a quite nice way to handle this situation, which I propose as a new feature for Linphone: have a custom button that automatically issues a configurable DTMF code, and place it prominently on the call screen. One tap takes care of everything.

This involves the following logic:

- on the account configuration, have a new setting called "Custom DTMF code", which can either be **disabled** or hold a value made of digits and `#`  and `*`. I would also be able to give it a text. So I would enter there: 
  - I want this **enabled**, 
  - and my custom code is **`7*`**
  - and the button text is **`Open`**

- on the "call ringing" and on the "call ongoing" screens:
  - for users who don't have it enabled, nothing shows, the screens stay the same as now
  - for users who have a code, a new button appears saying my text (or if the user left that blank, use "DTMF" or "CODE" or an appropriate icon as a default).
    - on the "call ringing" screen, this button **answers the call and sends the DTMF code**
    - on the "call ongoing" screen, this button just **sends the DTMF code** 
    - sending the code does not turn off the call, since sometimes a second attempt is needed

Suggested UI, mock-ups:

![image](https://user-images.githubusercontent.com/15945027/213734561-52809205-40c0-4181-bb16-512837cb5bf7.png)

![image](https://user-images.githubusercontent.com/15945027/213735089-ac5f8785-610a-43bf-bf3b-3ecb5ade2cd9.png)

(I'm sure you can find better placements and nicer looking graphics, but you get the general idea 😄 )

Note that this is not just "my" feature request. Lots of people use DTMF codes for millions of things. Of course a complicated UI with many programmable buttons could be designed, but I suspect that a single code that is repeatedly used makes up for 99% of the use cases.

## **3. Keyboard overlay while using DTMF**

Any way, on video calls this Numpad overaly **should be made smaller, or more transparent**, or both. I can't see if the people were able to open the door (it doesn't work very well) and came in, or if I have to do the code again. 

![image](https://user-images.githubusercontent.com/15945027/213736114-0c38cb51-b62f-4dc0-8abb-6db33d3b8b38.png)

I have to put down the keypad to see, and then go through all the taps again if I need to repeat the command...

## **4. Fully silent calls... Call volume versus media volume**

This one is a trickier feature request. Basically, I often answer the door while doing other things, talking to some co-worker, meetings, etc. I often don't need the audio dialog ("who is it?") at all, and I would appreciate my phone not making any noise (street noise, door buzzer). The video is good for this: I see "oh, it's John" and trigger the lock. And John comes in several times a day, so he knows the drill.

But it's very hard to do it silently because Linphone (correctly, for all the more common use cases) uses the Android "Call volume", which is impossible to mute. So even if my phone is in silent mode, the call generates audible noise when I answer it.

I would like the calls to start muted, and then have a button I could use to turn on sounds. I realize this should not be the default mode, since most people won't appreciate it for their ordinary SIP voice and video calls. 

But perhaps a new option in the account settings could provide this. Either come up with such a "silent mode" inside your app, or perhaps a better way to do it, while keeping things simple, is this:
- have a new account option to select either "Call volume" (this should be the default) or "Media volume" to be used during calls. 
- none of this affects "Ring volume".
- "Call volume" is the default, and makes things work like they currently do.
- But if I select "Media volume", Linphone connects the call using my _other_ Android audio stream (the one that is used when a video is playing, for example). This way I can use the phone's normal controls (and other apps like Tasker that manipulate it) to manage my silent mode easily, or turn up the volume during the call if I really need it.

## Motivation

These VoIP video door phones are getting sold in huge numbers. A single one can be used by 40 or 80 or 400 people in a large building. Linphone is currently, in my opinion, a great solution to use for this, and it is free, where others (like 2N's proprietary app) charge monthly fees! Imagine if Linphone was to be made not just good, but excellent, for this use case, standing out above the competition. It would quickly bring in thousands of new users, and generate a lot of brand recognition. Eventually this could translate into more licensing and support contracts as people decide to use it in their companies, at work, etc.

## Technical info

 - Device: OnePlus Nord CE 5G
 - OS: Android 11
 - Version of the App: [5.0.5](https://www.youtube.com/watch?v=iV5VKdcQOJE)
 - Version of the SDK: 5.2.15
 - Where you did got it from: Play Store
 - Oxygen OS 11.0.5

Thanks! Great work you're doing here! 🎉

**Comment by Viish** [comment]
Hi,

Thanks for taking the time to write such a list of improvements, and we're glad you like our softphone.

As you already noticed, linphone is primarily made for calls, not so much for home surveillance.
But for that we have another app, using the same SDK, that is called linhome: https://gitlab.linphone.org/BC/public/linhome-android
I believe it is very close to what you desire in linphone.

Cheers,

**Comment by pgorod** [comment]
Ok... that's a surprise. A nice surprise.

So, I've been trying to set up Linhome and it seems to do the job... almost!

I define an Action to "open door", code is "7*", but during the call I don't get any button to run the action. I've tried full-screen and normal view, I've tried tapping everywhere but I can't seem to find a way to execute the code...

Is this a bug, or is it me?

**Comment by Viish** [comment]
It may be a bug, or a missing feature as it may only currently support 1 DTMF command.
I'm not involved in linhome development and haven't tested it enough to answer you more precisely, sorry.
You may want to open an issue on the linhome repository so at least this info won't be forgotten.

**Comment by Viish** [comment]
The GitHub repo is https://github.com/BelledonneCommunications/linhome-android

**Comment by pgorod** [comment]
I opened [an issue](https://github.com/BelledonneCommunications/linhome-android/issues/30) there, but it hasn't been looked at yet. I hope that repo is still getting maintained, there don't seem to be many recent commits...

I've had more time to test Linhome and there are two "deal-breakers", stuff that really won't work:

- **not being able to send the DTMF code**, that feature seems to be broken
- **no way to enable/disable easily**. In Linphone I do it with "disable account", when I am away from work and I am not responsioble for answering the door; I don't want to be annoyed all day with this. In Linhome, I really don't see a way to disable it except disconnecting the account (or uninstalling the app!), but those steps are a bit too radical. Enabling again, after that, is a lot of work 😓 

I was wondering which parts of this Issue here are relevant to Linphone, even after knowing that there is a separate app that (supposedly) fills the needs I specified...:

- even if Linhome worked well, I am not sure I would prefer it; Linhpone is a good app, and it's more generic, so I can use it for any kind of SIP call.
- the idea of keeping video visible is probably a reasonable idea for any kind of video call. So, my number 1 request above surely makes sense. And a bit of shrinking the Numpad and a bit more transparency is easy to implement, and won't hurt, as explained in number 3.
- I can't evaluate how useful number 2 can be for other people. Which DTMF codes do people use regularly during SIP calls? Anyone use codes repeatedly, for example, to invoke a function on their company's PBX, like a transfer?
- number 4 adds flexibility to Linphone. It lets people handle volume issues with other apps and OS commands. I believe this adds value.

**Comment by Viish** [comment]
Hi,

Don't worry linhome is still being developed, a new release will be published soon.
I assigned your issue directly to the linhome développer so it won't be forgotten.

Regarding your linphone suggestions, your first point is indeed relevant. I believe it to be either a bug or simply something I forgot when I added the early media preview, I'll check it out for next release.
Points 2 & 3 won't be implemented as we used a UX/UI agency to create the new design and I can't make changes just like that. However we have a full UI rewrite coming this year and we'll keep that suggestion in our mind for the guidelines for the agency.
Finally for point 4 this isn't easily doable (at least the way you suggest it). What could be done is a mute button like on the desktop app that shuts off/on the received audio.

Cheers,

**Comment by Viish** [comment]
Hi again,

I just fixed your first issue in commit 1f92c637, will be part of next minor release.

Cheers,

**Comment by pgorod** [comment]
Great work, thanks.

I'll leave it to you to close this issue when you think it's appropriate - I don't if you want to leave it here while the rest of the issues evolve, or if you just prefer to clean up the issues list. Any way, it's ok with me. Thanks for your assistance.

**Comment by Viish** [comment]
We can keep this issue open if you'd like, it has the `enhancement` label so it won't be mishandled.

**Comment by Viish** [comment]
Hi,

From what I'm being told by our project manager, all dev is finished, we just have to do through the test plan before releasing it, so it should happen 'soon'.

Cheers,

**Comment by pergolafabio** [comment]
I see changes on gitlab for the linhome 1.2, but still not released, do you know when it's expected?

**Comment by Viish** [comment]
We are testing it, and once the test plan will be all green we will release it, shouldn't take long.

**Comment by pergolafabio** [comment]
Great, looking forward to it...

Is it maybe possible to share an APK already to test? I also subscribed in beta program on playstore, but there are none

**Comment by pergolafabio** [comment]
Or is there maybe some public ftp where we can download test builds?

**Comment by pgorod** [comment]
Yeah, anything to break this stalemate would be really useful. This delay has made me look like a fool quite a few times in the past months. I am relaying the promises of "soon" to other people, and I always come up empty-handed :-(

**Comment by pgorod** [comment]
@Viish can we get a test build of the new LinHome please? This continuous delay is causing me professional damage. I can't keep promising something and failing to deliver. I understand things sometimes get complicated in companies, and not everyone is moving forward, for one reason or another. 

But too many months have gone by. If you can just provide an apk I can install, then we can wait for the official release whenever your company is ready. I would really appreciate this. Thanks.

**Comment by pergolafabio** [comment]
same for me , i'm following the release here:

https://gitlab.linphone.org/BC/public/linhome-android/-/commits/release/1.2

but would be great if we have an APK already

**Comment by Viish** [comment]
Hi, release is undergoing test plan, you can use release/1.2 branch from repository.
If not yet available as public beta it will be soon.
I'm currently in holidays so I can tell you much more.

**Comment by pergolafabio** [comment]
Subscribed to beta, there is indeed a 10th July release available!

**Comment by pgorod** [comment]
@Viish thanks for your feedback, at least it's nice to know a little bit more. Have a great holiday, and sorry to bother you!

@pergolafabio I am not familiar with Gitlab and this project's deployments. Can you please elaborate? Does that release have an apk I can install? Or would it require some sort of compilation? If there is an apk for a Beta, how can I get it? Thanks!

**Comment by pgorod** [comment]
That's even better! Ok, I got the new Beta, and it works fine... but it still doesn't have the (very basic) feature I need, the ability to turn off the app (or the account, or the device...), I mean, some way to avoid being bothered when I am not on site. :-(

**Comment by pergolafabio** [comment]
maybe you can setup those rules on asterisk side ?

**Comment by pgorod** [comment]
It's coming directly from a 2N IP Verso. I can use the API to change the way the buttons are configured, so it is possible. But it is not trivial, especially if I want to do it for many users, some of them non-technical people. They should be able to just use a checkbox on their app...

**Comment by Viish** [comment]
As the Linhome release has been published a few weeks ago now I'm closing this issue.

**Comment by pgorod** [comment]
Thanks. I assume this is Linhome Release 1.2 we're talking about.

I wish the other repo for Linhome had some communication going on, it feels like a ghost town... 

I spent a week of development time to work around the fact that Linhome does not have a simple check-box to disable it when people don't want to take calls. I had to make a big complex system from the Home Assistant app we use, which in turn connects to FreePBX, so that the call does not go out to that user's phone when he has disabled it. Oh, and this also forced me to have personal SIP accounts for each person, when a generic one could have sufficed.

I still find Linhome the best free app out there for this purpose. It's a pity that its development is clearly a low priority for the company running it, but hey, that's life, I can't ask for much if I am getting the app for free.

Best wishes (sincerely)
Pedro
